File: src/main/java/utils/Crypto.java
package utils;
import logging.Logger;
import javax.crypto.Cipher;
import java.io.*;
import java.security.*;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

import static java.nio.charset.StandardCharsets.UTF_8;

public class Crypto {
    private KeyPair keyPair;
    private String keystorePath;

    public Crypto(String keystorePath){
        this.keystorePath = keystorePath;
        //try to read the keypair from local storage
        try {
            keyPair = loadKeyPair(keystorePath);
            Logger.INSTANCE.info("Loaded KeyPair from: " + keystorePath);
        } catch (IOException e) {
            //assume local storage non-existent or corrupted so generate new keypair
            KeyPairGenerator generator = null;
            try {
                generator = KeyPairGenerator.getInstance("RSA");
                generator.initialize(2048, new SecureRandom());
                keyPair = generator.generateKeyPair();
                Logger.INSTANCE.info("Generated new KeyPair");
                saveKeyPair(keystorePath,keyPair);
                Logger.INSTANCE.info("Saved KeyPair to: " + keystorePath);
            } catch (NoSuchAlgorithmException ex) {
                ex.printStackTrace();
            } catch (IOException ex) {
                Logger.INSTANCE.error("Failed saving KeyPair to: " + keystorePath);
                //ex.printStackTrace(); //non breaking
            }
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch (InvalidKeySpecException e) {
            e.printStackTrace();
        }
    }

    public static String encrypt(String plainText, PublicKey publicKey) throws Exception {
        Cipher encryptCipher = Cipher.getInstance("RSA");
        encryptCipher.init(Cipher.ENCRYPT_MODE, publicKey);

        byte[] cipherText = encryptCipher.doFinal(plainText.getBytes(UTF_8));

        return Base64.getEncoder().encodeToString(cipherText);
    }

    public static String decrypt(String cipherText, PrivateKey privateKey) throws Exception {
        byte[] bytes = Base64.getDecoder().decode(cipherText);

        Cipher decriptCipher = Cipher.getInstance("RSA");
        decriptCipher.init(Cipher.DECRYPT_MODE, privateKey);

        return new String(decriptCipher.doFinal(bytes), UTF_8);
    }

    public static String sign(String plainText, PrivateKey privateKey) throws Exception {
        Signature privateSignature = Signature.getInstance("SHA256withRSA");
        privateSignature.initSign(privateKey);
        privateSignature.update(plainText.getBytes(UTF_8));

        byte[] signature = privateSignature.sign();

        return Base64.getEncoder().encodeToString(signature);
    }

    public static boolean verify(String plainText, String signature, String publicKey) throws Exception {
        //BASE64Decoder decoder = new BASE64Decoder();
        //byte[] byteKey = decoder.decodeBuffer(publicKey);
        byte[] byteKey = java.util.Base64.getMimeDecoder().decode(publicKey);
        Signature publicSignature = Signature.getInstance("SHA256withRSA");
        X509EncodedKeySpec X509publicKey = new X509EncodedKeySpec(byteKey);
        KeyFactory kf = KeyFactory.getInstance("RSA");
        publicSignature.initVerify(kf.generatePublic(X509publicKey));
        publicSignature.update(plainText.getBytes(UTF_8));

        byte[] signatureBytes = Base64.getDecoder().decode(signature);

        return publicSignature.verify(signatureBytes);
    }
    public void saveKeyPair(String path, KeyPair keyPair) throws IOException {
        PrivateKey privateKey = keyPair.getPrivate();
        PublicKey publicKey = keyPair.getPublic();

        // Store Public Key.
        X509EncodedKeySpec x509EncodedKeySpec = new X509EncodedKeySpec(
                publicKey.getEncoded());
        FileOutputStream fos = new FileOutputStream(path + "/public.key");
        fos.write(x509EncodedKeySpec.getEncoded());
        fos.close();

        // Store Private Key.
        PKCS8EncodedKeySpec pkcs8EncodedKeySpec = new PKCS8EncodedKeySpec(
                privateKey.getEncoded());
        fos = new FileOutputStream(path + "/private.key");
        fos.write(pkcs8EncodedKeySpec.getEncoded());
        fos.close();
    }

    public KeyPair loadKeyPair(String path) throws IOException, NoSuchAlgorithmException,
        InvalidKeySpecException {
        // Read Public Key.
        File filePublicKey = new File(path + "/public.key");
        FileInputStream fis = new FileInputStream(path + "/public.key");
        byte[] encodedPublicKey = new byte[(int) filePublicKey.length()];
        fis.read(encodedPublicKey);
        fis.close();

        // Read Private Key.
        File filePrivateKey = new File(path + "/private.key");
        fis = new FileInputStream(path + "/private.key");
        byte[] encodedPrivateKey = new byte[(int) filePrivateKey.length()];
        fis.read(encodedPrivateKey);
        fis.close();

        // Generate KeyPair.
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        X509EncodedKeySpec publicKeySpec = new X509EncodedKeySpec(
                encodedPublicKey);
        PublicKey publicKey = keyFactory.generatePublic(publicKeySpec);

        PKCS8EncodedKeySpec privateKeySpec = new PKCS8EncodedKeySpec(
                encodedPrivateKey);
        PrivateKey privateKey = keyFactory.generatePrivate(privateKeySpec);

        return new KeyPair(publicKey, privateKey);
    }

    public KeyPair getKeyPair() {
        return keyPair;
    }

    public String getPublicKey(){
        Key pubKey = keyPair.getPublic();
        return new String(java.util.Base64.getMimeEncoder().encode(pubKey.getEncoded()));
    }
}

File: src/main/kotlin/utils/Crypto.kt
package utils

import logging.Logger
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.security.*
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.*

class Crypto(private val keystorePath: String) {

    // TODO make immutable, fix public key property.

    private var keyPair: KeyPair
    var encodedPublicKey: ByteArray
        private set
    var publicKey: String
        private set
    var hashedPublicKey: ByteArray
        private set

    /** Signs data using the keypair from [keystorePath]. */
    @Throws(Exception::class)
    fun sign(data: ByteArray?): ByteArray {
        val privateSignature = Signature.getInstance("SHA256withRSA")
        privateSignature.initSign(keyPair.private)
        privateSignature.update(data)
        return privateSignature.sign()
    }

    /** Verifies whether the data signed by the public key equals to the signature. */
    @Throws(Exception::class)
    fun verify(data: ByteArray?, signature: ByteArray?, publicKey: String?): Boolean {
        val byteKey = Base64.getMimeDecoder().decode(publicKey)
        val publicSignature = Signature.getInstance("SHA256withRSA")
        val X509publicKey = X509EncodedKeySpec(byteKey)
        val kf = KeyFactory.getInstance("RSA")
        publicSignature.initVerify(kf.generatePublic(X509publicKey))
        publicSignature.update(data)
        return publicSignature.verify(signature)
    }

    /** Stores the [keyPair] in [path] */
    @Throws(IOException::class)
    fun saveKeyPair(path: String, keyPair: KeyPair) {
        val privateKey = keyPair.private
        val publicKey = keyPair.public

        // Store Public Key.
        val x509EncodedKeySpec = X509EncodedKeySpec(
            publicKey.encoded
        )
        var fos = FileOutputStream("$path/public.key")
        fos.write(x509EncodedKeySpec.encoded)
        fos.close()

        // Store Private Key.
        val pkcs8EncodedKeySpec = PKCS8EncodedKeySpec(
            privateKey.encoded
        )
        fos = FileOutputStream("$path/private.key")
        fos.write(pkcs8EncodedKeySpec.encoded)
        fos.close()
    }

    /** Loads and returns the keypair stored in [path] or throws an exception if the pair does not exist. */
    @Throws(Exception::class)
    fun loadKeyPair(path: String): KeyPair {
        // Read Public Key.
        val filePublicKey = File("$path/public.key")
        var fis = FileInputStream("$path/public.key")
        val encodedPublicKey = ByteArray(filePublicKey.length().toInt())
        fis.read(encodedPublicKey)
        fis.close()

        // Read Private Key.
        val filePrivateKey = File("$path/private.key")
        fis = FileInputStream("$path/private.key")
        val encodedPrivateKey = ByteArray(filePrivateKey.length().toInt())
        fis.read(encodedPrivateKey)
        fis.close()

        // Generate KeyPair.
        val keyFactory = KeyFactory.getInstance("RSA")
        val publicKeySpec = X509EncodedKeySpec(encodedPublicKey)
        val publicKey = keyFactory.generatePublic(publicKeySpec)
        val privateKeySpec = PKCS8EncodedKeySpec(encodedPrivateKey)
        val privateKey = keyFactory.generatePrivate(privateKeySpec)
        return KeyPair(publicKey, privateKey)
    }

    init {
        keyPair = try {
            loadKeyPair(keystorePath)
        } catch (e: Exception) {
            Logger.error("Could not load key pair: $e.")
            Logger.info("Generating new key pair.")
            val generatedPair = KeyPairGenerator.getInstance("RSA").let {
                it.initialize(2048, SecureRandom())
                it.generateKeyPair()
            }
            // saveKeyPair(keystorePath, generatedPair)
            generatedPair
        }
        encodedPublicKey = keyPair.public.encoded
        hashedPublicKey = Utils.sha256(encodedPublicKey)
        publicKey = String(Base64.getMimeEncoder().encode(encodedPublicKey))
    }
}